crates/pera-framework/packages/pera-system/sources/dwallet_2pc_mpc_ecdsa_k1.move

```diff
@@ -865,4 +922,121 @@ module pera_system::dwallet_2pc_mpc_ecdsa_k1 {
+            transfer::transfer(presign, SYSTEM_ADDRESS);
+            i = i + 1;
+        };
+        presigns_bytes.reverse();
+        presign_session_ids.reverse();
+        presigns.destroy_empty();
+        assert!(
+            verify_partially_signed_signatures_native(
+                signatures,
+                messages,
+                presigns_bytes,
+                get_dwallet_output(dwallet)
+            ),
+            EInvalidSignatures
+        );
+        let partial_signatures = PartiallySignedMessages {
+            id: object::new(ctx),
+            presigns: presigns_bytes,
+            presign_session_ids,
+            messages,
+            signatures,
+            dwallet_output: get_dwallet_output(dwallet),
+            dwallet_id: object::id(dwallet),
+            dwallet_cap_id: get_dwallet_cap_id(dwallet),
+        };
+        event::emit(CreatedPartiallySignedMessagesEvent {
+            partial_signatures_object_id: object::id(&partial_signatures),
+        });
+        transfer::transfer(partial_signatures, tx_context::sender(ctx));
+    }
+
+    /// A function to launch a sign flow with a previously published [`PartiallySignedMessages`].
+    ///
+    /// See the docs of [`PartiallySignedMessages`] for more details on when this may be used.
+    public fun future_sign(
+        partial_signature: PartiallySignedMessages,
+        message_approvals: &mut vector<MessageApproval>,
+        ctx: &mut TxContext
+    ) {
+        let PartiallySignedMessages {
+            id,
+            mut presigns,
+            mut presign_session_ids,
+            mut messages,
+            mut signatures,
+            dwallet_id,
+            dwallet_cap_id,
+            dwallet_output,
+        } = partial_signature;
+        object::delete(id);
+        let message_approvals_len = vector::length(message_approvals);
+        let messages_len = vector::length(&messages);
+        assert!(message_approvals_len == messages_len, EApprovalsAndMessagesLenMismatch);
+        let batch_session_id = object::id_from_address(tx_context::fresh_object_address(ctx));
+        event::emit(StartBatchedSignEvent {
+            session_id: batch_session_id,
+            hashed_messages: messages,
+            initiator: tx_context::sender(ctx)
+        });
+        let mut i = 0;
+        while (i < message_approvals_len) {
+            let message_approval = vector::pop_back(message_approvals);
+            let (message_approval_dwallet_cap_id, approved_message) = remove_message_approval(message_approval);
+            assert!(dwallet_cap_id == message_approval_dwallet_cap_id, EMesssageApprovalDWalletMismatch);
+            let message = vector::pop_back(&mut messages);
+            assert!(&message == &approved_message, EMissingApprovalOrWorngApprovalOrder);
+            let id = object::id_from_address(tx_context::fresh_object_address(ctx));
+            let centralized_signed_message = vector::pop_back(&mut signatures);
```

Comment:
It seems the code here is repeated in lines 798-... Consider extracting it into a separate function. If that's not possible, add line separations to improve readability.